### PR TITLE
feat: allow adding Indexes for columns

### PR DESCRIPTION
**Pull request by singh1114**
Adds new property name index

Signed-off-by: Ranvir Singh <ranvir.singh1114@gmail.com>

### CODE REVIEW

**Comment by singh1114** [comment]
**Tests**

- [ ] Check if new tables are created with indexes.
- [ ] Check if indexes are created properly to old tables.
- [ ] Check if when indexes are removed from code, indexes are removed properly.

**Comment by B4nan** [comment]
Please base this on dev branch. The method you updated does not even exist in the latest version.

File: lib/types.ts [hunk 1]
```diff
@@ -129,6 +129,7 @@ export interface EntityProperty<T extends AnyEntity<T> = any> {
   inverseJoinColumn: string;
   referenceColumnName: string;
   referencedTableName: string;
+  index?: boolean;
```
Comment (B4nan) [inline lib/types.ts]: i guess this should be also `boolean | string`

File: tests/SchemaGenerator.test.ts [hunk 2]
```diff
@@ -201,6 +202,7 @@ describe('SchemaGenerator', () => {
           type: 'string',
           fieldName: 'name',
           columnType: 'varchar(255)',
+          index: true,
```
Comment (B4nan) [inline tests/SchemaGenerator.test.ts]: lets test the custom index name here too

File: lib/types.ts [hunk 3]
```diff
@@ -152,4 +152,5 @@ export interface EntityMetadata<T extends AnyEntity<T> = any> {
   customRepository: () => Constructor<EntityRepository<T>>;
   hooks: Partial<Record<HookType, (string & keyof T)[]>>;
   prototype: T;
+  index: boolean | string;
```
Comment (B4nan) [inline lib/types.ts]: index on entity metadata level needs to be array of possible indexes, also to support compound indexes it needs to be more than just an index name, you will have to have array of property names there too.

File: lib/decorators/Indexed.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import { EntityProperty, AnyEntity } from '../types';
+import { Utils } from '../utils';
+
+export function Index(options: IndexOptions = {}): Function {
+    return function (target: AnyEntity, propertyName: string) {
+        const meta = MetadataStorage.getMetadata(target.constructor.name);
+        Utils.lookupPathFromDecorator(meta);
+        const indexName = Utils.isString(options.name) ? options.name : true;
+        options.name = propertyName;
+        meta.index = options.name;
```
[the remark this answers is not in the document]
Comment (B4nan) [reply lib/decorators/Indexed.ts]: this is what should ultimately work:

```typescript
@Entity()
@Index({ properties: ['name', 'age'] }) // compound index, with generated name
@Index({ name: 'asd', properties: ['name'] }) // simple index, with custom name
export class Author {

  @Property()
  @Index() // generated name
  age?: number;

  @Index('born_index') // custom name, should be possible to pass it like this, as well as via options object like `@Index({ name: 'born_index' })`
  @Property()
  born?: Date;

}
```

[the remark this answers is not in the document]
Comment (B4nan) [reply lib/decorators/Indexed.ts]: also, this: 

```typescript
@Index({ name: 'asd', properties: ['name', 'age'] })
```

should be possible to write like this (I'd like to have both ways possible):

```typescript
@Index('asd', ['name', 'age'])
```

File: lib/decorators/index.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 export * from './PrimaryKey';
 export * from './SerializedPrimaryKey';
 export * from './Entity';
+export * from './Indexed';
```
[the remark this answers is not in the document]
Comment (singh1114) [reply lib/decorators/index.ts]: `index.ts` already exists editor doesn't allow me to `Index.ts` again.

File: lib/decorators/Indexed.ts [hunk 6]
```diff
@@ -9,11 +9,13 @@ export function Index(options: IndexOptions = {}): Function {
         const indexName = Utils.isString(options.name) ? options.name : true;
         options.name = propertyName;
         meta.index = options.name;
-        meta.properties[propertyName] = Object.assign({ index: indexName }, options) as EntityProperty;
+        meta.properties[propertyName] = Object.assign({
+            ...meta.properties[propertyName],
```
Comment (B4nan) [inline lib/decorators/Indexed.ts]: `meta.properties[propertyName]` can be possibly undefined (depends on the order of decorators), so I guess this should be:

```
...(meta.properties[propertyName] || {}),
```

File: lib/decorators/Property.ts [hunk 7]
```diff
@@ -9,7 +9,9 @@ export function Property(options: PropertyOptions = {}): Function {
     const meta = MetadataStorage.getMetadata(target.constructor.name);
     Utils.lookupPathFromDecorator(meta);
     options.name = options.name || propertyName;
-    const prop = Object.assign({ reference: ReferenceType.SCALAR }, options) as EntityProperty;
+    const prop = Object.assign({ ...meta.properties[propertyName],
```
Comment (B4nan) [inline lib/decorators/Property.ts]: put it on new line

```
    const prop = Object.assign({
      ...(meta.properties[propertyName] || {}),
      reference: ReferenceType.SCALAR,
    }, options) as EntityProperty;
```

File: lib/decorators/Property.ts [hunk 8]
```diff
@@ -20,7 +22,6 @@ export function Property(options: PropertyOptions = {}): Function {
       prop.type = 'method';
       prop.getterName = propertyName;
     }
-
```
Comment (B4nan) [inline lib/decorators/Property.ts]: keep this new line

File: tests/SchemaGenerator.test.ts (hunk 2 above)
Comment (singh1114) [reply tests/SchemaGenerator.test.ts]: Let's keep a test for passing a boolean as well.

**Comment by B4nan** [comment]
Where are we with this?

**Comment by B4nan** [comment]
I will close this one now. Please open new PR so we can get rid of that messed up history (161 commits in this). Also please let me know if you do not want to continue working on it, I will be happy to take over. What is in this PR is quite mergeable, I would be also open to merging not the full implementation I asked for, but want to see 100% code coverage and also updated docs.

**Comment by ryankauk** [comment]
Hey was wondering the progress on this? Considering switching to mikro but realized it was missing this.

**Comment by B4nan** [comment]
I am planning to support this in one of upcoming feature releases (got most of it implemented somewhere already). Will start with SQL implementation only, mongo support will follow probably in another feature release (together with support for transactions in mongo).

**Comment by ryankauk** [comment]
That's great, fine for me I'm working with SQL. Thanks for the prompt response. Is there a rough time estimation that you could give?

**Comment by B4nan** [comment]
Currently I have few things in progress/proof of concept stage, this being one of it. I will try to focus on that so it is out during few weeks.

**Comment by ryankauk** [comment]
No problem, thank you.

**Pull request closed without merging**